chore(doc): how to return multiple results from a function

nulltoken opened the pull request:
Fixes #1030 and #920 

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

This attempts at documenting how to properly return multiple results from custom functions while playing nicely with the deduplication of the fringerprinting mechanism.

nulltoken commented:
@philsturgeon I'd be delighted to benefit from your writing skills and help turn this bunch of words into something worthy of being merged. 🙏

@P0lip I've tried to be extra careful with the code as it might become a source for copy/paste. Could you please unleash your review magic on this?

philsturgeon commented on docs/guides/custom-functions.md:
```diff
@@ -145,6 +145,68 @@ rules:
 
+## Returning multiple results
+
+Spectral results are automatically deduplicated when they bear the same `code` and target the same `path`.
```
Many functions will return a single message, but its possible for a function to return multiple. 

For example, if a rule is created to make sure something is unique, it could either:

a) return a single error for the entire array which lists offending values in a comma separated list
b) return a single error for the array value which contains the first offending non-unique item
c) return multiple errors for each duplicate value located

How exactly you chose to implement messages depends on the rule at hand and probably personal preference too.

It's worth keeping in mind, Spectral will attempt to deduplicate messages based on....

nulltoken replied:
Fixed

P0lip commented on docs/guides/custom-functions.md:
```diff
@@ -145,6 +145,68 @@ rules:
+    for (let i = 0; i < targetVal.length; i++) {
+        const tagName = targetVal[i][NAME_PROPERTY];
+
+        if (!tagName) {
```
```suggestion
        if (tagName === void 0) {
```
// just in case tagName is an empty string.

nulltoken replied:
Fixed

P0lip commented on docs/guides/custom-functions.md:
```diff
@@ -145,6 +145,68 @@ rules:
+    const seen = [];
+    const results = [];
+
+    const rootPath = typeof (paths.target) !== 'undefined' ? paths.target : paths.given;
```
```suggestion
    const rootPath = paths.target !== void 0 ? paths.target : paths.given;
```
minor, feel free to ignore

nulltoken replied:
Fixed

P0lip commented on docs/guides/custom-functions.md:
```diff
@@ -145,6 +145,68 @@ rules:
+    const rootPath = typeof (paths.target) !== 'undefined' ? paths.target : paths.given;
+
+    for (let i = 0; i < targetVal.length; i++) {
+        const tagName = targetVal[i][NAME_PROPERTY];
```
```suggestion
        if (targetVal[i] === null || typeof targetVal[i] !== 'object') {
           continue;
        }

        const tagName = targetVal[i][NAME_PROPERTY];
```

nulltoken replied:
Fixed

nulltoken replied:
Fixed

P0lip commented on docs/guides/custom-functions.md:
```diff
@@ -145,6 +145,68 @@ rules:
+                },
+            );
+
+            console.error(results[results.length - 1]);
```
```suggestion
```

P0lip replied:
Let's not log anything.

nulltoken replied:
Duh... A forgotten witness of my crazy custom function debugging skills ;-)

Fixed

P0lip commented on docs/guides/custom-functions.md:
```diff
@@ -145,6 +145,68 @@ rules:
+            );
+
+            console.error(results[results.length - 1]);
+            continue;
```
guess we could just do `if/else`?
```
if (seen.includes(tagName)) {
  ...
} else {
  seen.push(tagName);
}
```
I believe this is slightly more readable, but as always, it's subjective.

nulltoken replied:
Fixed

P0lip commented on docs/guides/custom-functions.md:
````diff
@@ -97,7 +97,7 @@ You can do it as follows:
 ```yaml
 functions:
-- - equals
+- equals
````
Good catch.

nulltoken replied:
:)

P0lip commented on docs/guides/custom-functions.md:
```diff
@@ -145,6 +145,68 @@ rules:
+const NAME_PROPERTY = 'name';
+
+module.exports = (targetVal, _opts, paths) => {
+
```
```suggestion
```

nulltoken replied:
Fixed

nulltoken commented:
@philsturgeon @P0lip Thanks for the amazing feedback. I believe I've fixed all the outstanding comments.

I've also added a harness scenario to cover this.

P0lip commented:
@philsturgeon wanna add anything before we merge it?
